pyccel/ast/builtin_objects/list_functions.py

```diff
@@ -0,0 +1,79 @@
+# coding: utf-8
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
+#------------------------------------------------------------------------------------------#
+"""
+The List container has a number of built-in methods that are 
+always available.
+
+This module contains objects which describe these methods within Pyccel's AST.
+"""
+
+from pyccel.ast.datatypes import NativeVoid, NativeGeneric, NativeHomogeneousList
+from pyccel.ast.internals import PyccelInternalFunction
+
+
+class ListAppend(PyccelInternalFunction):
+    """
+    Represents a call to the .append() method.
+
+    Represents a call to the .append() method of an object with a list type,
+    which adds an element to the end of the list. This method returns `None`.
+    The append method is called as follows:
+
+    >>> a = [1]
+    >>> a.append(2)
+    >>> print(a)
+    [1, 2]
+
+    Parameters
+    ----------
+    lst_bound_arg : TypedAstNode
+        The variable representing the list.
+    
+    new_elem : TypedAstNode
+        The argument passed to append() method.
+    """
+    __slots__ = ("_lst_bound_arg", "_append_arg")
+    _dtype = NativeVoid()
+    _shape = None
+    _order = None
+    _rank = 0
+    _precision = -1
+    _class_type = NativeHomogeneousList()
+    name = 'append'
+
+    def __init__(self, lst_bound_arg, new_elem) -> None:
+        check = (getattr(lst_bound_arg, 'order', None) == getattr(new_elem, 'order', None))
+        conditions = (
+            new_elem.dtype is not NativeGeneric() and
+            lst_bound_arg.dtype == new_elem.dtype and
+            lst_bound_arg.precision == new_elem.precision and
+            lst_bound_arg.rank - 1 == new_elem.rank and
+            (check if check else (lst_bound_arg.class_type == new_elem.class_type))
+        )
+        is_homogeneous = lst_bound_arg.dtype is not NativeGeneric() and conditions
+        if not is_homogeneous:
+            raise TypeError("Expecting an argument of the same type as the elements of the list")
+        self._lst_bound_arg = lst_bound_arg.name
+        self._append_arg = new_elem
+        super().__init__()
+
+    @property
+    def list_variable(self):
+        """
+        Get the variable name representing the list.
+
+        Get the variable name representing the list.
```

Comment:
According to this docstring the method returns a name (a string). Is this correct? I would expect it to return a `Variable`.

Reply:
yes that's correct, it should return the name of the list which is a string, i'll update the method name

Reply:
That sounds like a problem. That means you can't append to a list stored in a class right?

Reply:
i didn't quite understand what you meant by "**That means you can't append to a list stored in a class**", can you elaborate more please ?

Reply:
The name of a class property depends on the language. For example:
```python
class A:
    def __init__(self, elem : int):
        self.my_list : list[int] = []
        self.my_list.append(elem)
```
Inside Pyccel `lst_bound_arg.name` will call `DottedVariable.name` this returns `my_list`. But when printing we want `self.my_list`. You can get this with `str(lst_bound_arg)` but this will still cause problems. The C translation would be something like:
```C
void A__init__(A self, int64_t elem):
    create(self.my_list)
    append(self.my_list, elem)
```
But the Fortran translation would be something like:
```fortran
subroutine A__create(self, elem)
    type(A), intent(inout) :: self
    integer(I64), intent(inout) :: elem

    create(self % my_list)
    append(self % my_list, elem)
end subroutine A__create
```
This is why the AST should never take care of deciding how something should be printed